.eslintrc.js

```diff
@@ -69,6 +73,131 @@ module.exports = {
                     message: "`assert.doesNotThrow()` should be replaced with a comment next to the code."
                 }]
             }
+        },
+
+        // Restrict relative path imports
+        {
```

Comment:
I don't think those needs to be checked in. They are good for refactoring, but once it's merged, there shouldn't be any problems with incorrect imports.

Reply:
I like the idea of checking some of these in. We do have [architecture guidelines](https://eslint.org/docs/developer-guide/architecture) specifying which files should or shouldn't use certain modules, and it's always nice to enforce those conventions via linting where possible.

Reply:
Thank you for the feedback.

I wanted to keep the following things in the future bug fixes and enhancements in order to keep the loose coupling:

- prevent making circular dependencies between those directories.
- minimizing the internal files which are depended by other directory's files.

I added a rough dependency graph into [architecture.md](https://github.com/eslint/eslint/blob/695225c166e27d7a59a41afbeb5ab3e9f4aa575c/docs/developer-guide/architecture.md).

@platinumazure Thank you for pointing out. I updated `architecture.md` with the new file structure.